src/common/transformations/src/transformations/common_optimizations/eliminate_loop_inputs_outputs.cpp

```diff
@@ -0,0 +1,303 @@
+#include "openvino/core/type.hpp"
+#include "openvino/op/loop.hpp"
+#include "openvino/op/parameter.hpp"
+#include "openvino/op/result.hpp"
+#include "openvino/op/tensor_iterator.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+
+using namespace std;
+using namespace ov;
+using namespace ov::op;
+using namespace ov::op::util;
+using namespace ov::element;
+using namespace ov::pass::pattern;
+
+using InvariantD = MultiSubGraphOp::InvariantInputDescription;
+using SlicedD = MultiSubGraphOp::SliceInputDescription;
+using MergedD = MultiSubGraphOp::MergedInputDescription;
+using OutputD = MultiSubGraphOp::BodyOutputDescription;
+using ConcatD = MultiSubGraphOp::ConcatOutputDescription;
+
+using ResultPtr = std::shared_ptr<v0::Result>;
+using OutputDescPtr = MultiSubGraphOp::OutputDescription::Ptr;
+using OutputDescMap = std::unordered_map<ResultPtr, OutputDescPtr>;
+using InputDescPtr = MultiSubGraphOp::InputDescription::Ptr;
+using BodyResultIdxMap = std::unordered_map<ResultPtr, uint64_t>;
+
+namespace {
+std::unordered_set<uint64_t> remove_results(const std::shared_ptr<v0::Parameter>& param,
+                                            const InputDescPtr& input_desc,
+                                            OutputDescMap& output_desc_map,
+                                            const BodyResultIdxMap& result_map,
+                                            int64_t current_iteration_input_idx,
+                                            int64_t body_condition_output_idx) {
+    if (current_iteration_input_idx == static_cast<int64_t>(input_desc->m_body_parameter_index)) {
+        // skip Parameter node storing the current iteration
+        return {};
+    }
+
+    // remove sub-graphs when Parameter is going to only Result nodes
+    // but Parameter node must be marked as Invariant or Merged Parameter
+    // and Result node must be unmarked (no OutputDescription) or marked as BodyOutputDescription
+    // all other cases do not guarantee to leave a graph valid
+    if (!ov::as_type_ptr<InvariantD>(input_desc) && !ov::as_type_ptr<MergedD>(input_desc)) {
+        return {};
+    }
+
+    std::unordered_set<uint64_t> removed_result_inds;
+    for (const auto& target_input : param->get_output_target_inputs(0)) {
+        const auto& consumer = target_input.get_node()->shared_from_this();
+        const auto& result = ov::as_type_ptr<v0::Result>(consumer);
+        if (!result) {
+            continue;
+        }
+        if (static_cast<int64_t>(result_map.at(result)) == body_condition_output_idx) {
+            // body condition output and related Parameter node must not be removed
+            continue;
+        }
+
+        const auto& output_desc = output_desc_map[result];
+        if (output_desc && ov::as_type_ptr<OutputD>(output_desc)) {
+            removed_result_inds.insert(result_map.at(result));
+        } else if (!output_desc) {
+            removed_result_inds.insert(result_map.at(result));
```

Comment:
no difference between 1st and 2nd if branches 
could you elaborate on why do we need them?

Reply:
Fixed.